test/unit/cbus_call.c

```diff
@@ -0,0 +1,178 @@
+#include "exception.h"
+#include "memory.h"
+#include "cbus.h"
+#include "unit.h"
+
+static struct fiber *caller_fiber;
+static struct cpipe pipe_to_callee;
+static struct cpipe pipe_to_caller;
+
+static int
+func(struct cbus_call_msg *msg)
+{
+	usleep(100000);
+	return 0;
+}
+
+/** Check ordinary cbus_call, nothing special. */
+static void
+test1(void)
+{
+	struct cbus_call_msg msg;
+	int rc = cbus_call(&pipe_to_callee, &pipe_to_caller, &msg, func, NULL,
+			   TIMEOUT_INFINITY);
+	is(rc, 0, "cbus_call ordinary");
+}
+
+static int
+empty(struct cbus_call_msg *msg)
+{
+	return 0;
+}
+
+/** Block until previously called func have been completed. */
+static void
+barrier(void)
+{
+	struct cbus_call_msg msg;
+	int rc = cbus_call(&pipe_to_callee, &pipe_to_caller, &msg, empty, NULL,
+			   TIMEOUT_INFINITY);
+	fail_if(rc != 0);
+}
+
+/** Set cbus_call timeout to 10 ms, while func runs for 100 ms. */
+static void
+test2(void)
+{
+	struct cbus_call_msg msg;
+	int rc = cbus_call(&pipe_to_callee, &pipe_to_caller, &msg, func, NULL,
+			   0.01);
+	struct error *err = diag_last_error(diag_get());
+	bool pass = (rc == -1) && err && (err->type == &type_TimedOut);
+	ok(pass, "cbus_call timeout");
+	barrier();
+}
+
+static int
+waker_fn(va_list ap)
+{
+	fiber_sleep(0.05);
+	fiber_wakeup(caller_fiber);
+	return 0;
+}
+
+/** Check that cbus_call is not interrupted by fiber_wakeup. */
+static void
+test3(void)
+{
+	struct fiber *waker_fiber = fiber_new("waker", waker_fn);
+	fail_if(waker_fiber == NULL);
+	fiber_wakeup(waker_fiber);
+
+	struct cbus_call_msg msg;
+	int rc = cbus_call(&pipe_to_callee, &pipe_to_caller, &msg, func, NULL,
+			   TIMEOUT_INFINITY);
+	is(rc, 0, "cbus_call wakeup");
+	barrier();
+}
+
+static int
+canceler_fn(va_list ap)
+{
+	fiber_sleep(0.05);
+	fiber_cancel(caller_fiber);
+	return 0;
+}
+
+/** Check that cbus_call is not interrupted by fiber_cancel. */
+static void
+test4(void)
```

Comment:
Please name the test cases appropriately. For example, `test_timeout`, `test_cancel`.

Reply:
Done.